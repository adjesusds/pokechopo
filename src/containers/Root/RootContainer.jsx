// External Packages
import { bool, func, number, shape, string } from 'prop-types'
import React, { Fragment } from 'react'
import { connect } from 'react-redux'

// Internal Packages
import { fetchPokemon } from 'actions/pokemon'
import { Button, Copyright, Divider, Link, Loading, Message } from 'components'
import constants from 'config/constants'

/**
 * Creates a new Root container.
 * @class
 */
const RootContainer = ({ data, errorMessage, isFetching, onDiscoverPokemonClick }) => (
  <Fragment>
    <Button onClick={onDiscoverPokemonClick} disabled={isFetching}>
      Discover a Pok√©mon
    </Button>
    <Divider />
    {!data ? (
      <Fragment>
        <Message variant="warning">{!isFetching ? 'Click it.' : <Loading message="Searching" />}</Message>
        {!errorMessage ? (
          <Message>
            Gotta catch &#39;em all!<Message variant="info">&trade;</Message>
          </Message>
        ) : (
          <Message variant="danger">{errorMessage}</Message>
        )}
      </Fragment>
    ) : (
      <Fragment>
        <Message>
          <Link href={`https://www.pokemon.com/us/pokedex/${data.name}/`} target="blank">
            {`${data.name[0].toUpperCase()}${data.name.slice(1)}`}
          </Link>
          <Message variant="warning">
            {` #${String(data.id).padStart(String(constants.POKEMON_MAX_NUMBER).length, '0')}`}
          </Message>
        </Message>
        <Message variant="success">Look at what you got!</Message>
      </Fragment>
    )}
    <Copyright>
      &copy; {new Date().getFullYear()}{' '}
      <Link href="https://www.linkedin.com/in/relztic/" target="blank">
        Relztic
      </Link>{' '}
      üê¥
    </Copyright>
  </Fragment>
)

/** @override */
RootContainer.propTypes = {
  data: shape({
    id: number.isRequired,
    name: string.isRequired,
  }),
  errorMessage: string.isRequired,
  isFetching: bool.isRequired,
  onDiscoverPokemonClick: func.isRequired,
}

/** @override */
RootContainer.defaultProps = {
  data: null,
}

/** @override */
const mapStateToProps = state => state.pokemon

/** @override */
const mapDispatchToProps = dispatch => ({
  onDiscoverPokemonClick: event => {
    event.preventDefault()

    const pokemonNumber = Math.floor(Math.random() * constants.POKEMON_MAX_NUMBER) + constants.POKEMON_MIN_NUMBER

    dispatch(fetchPokemon(pokemonNumber))
  },
})

/**
 * @constant {Element} ConnectedRootContainer The connected Root container, Redux.
 */
const ConnectedRootContainer = connect(mapStateToProps, mapDispatchToProps)(RootContainer)

export default ConnectedRootContainer

export { RootContainer as PlainRootContainer, mapDispatchToProps, mapStateToProps }
