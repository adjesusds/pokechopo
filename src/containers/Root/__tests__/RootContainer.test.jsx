// External Packages
import { expect as chaiExpect } from 'chai'
import { shallow } from 'enzyme'
import { reset, set } from 'mockdate'
import React from 'react'
import { spy } from 'sinon'

// Internal Packages
import { PlainRootContainer as RootContainer, mapDispatchToProps, mapStateToProps } from '../RootContainer'

/**
 * @constant {Object} state The state to mock.
 */
const state = {
  pokemon: {
    id: 1,
    name: 'bulbasaur',
  },
}

describe('<RootContainer />', () => {
  beforeAll(() => {
    set(new Date(2018, 0))
  })

  afterAll(() => {
    reset()
  })

  it('should render the container on valid props', () => {
    let container

    container = shallow(
      <RootContainer data={null} errorMessage="" isFetching={false} onDiscoverPokemonClick={() => {}} />
    )

    expect(container).toMatchSnapshot()

    container = shallow(<RootContainer data={null} errorMessage="" isFetching onDiscoverPokemonClick={() => {}} />)

    expect(container).toMatchSnapshot()

    container = shallow(
      <RootContainer
        data={null}
        errorMessage="InvalidArgumentError: invalid pokemonNumber `-1`"
        isFetching={false}
        onDiscoverPokemonClick={() => {}}
      />
    )

    expect(container).toMatchSnapshot()

    container = shallow(
      <RootContainer data={state.pokemon} errorMessage="" isFetching={false} onDiscoverPokemonClick={() => {}} />
    )

    expect(container).toMatchSnapshot()
  })

  it('should map the state on valid states', () => {
    const received = mapStateToProps(state)

    const expected = state.pokemon

    expect(received).toEqual(expected)
  })

  it('should map the dispatch on valid dispatches', () => {
    const dispatch = spy()

    const received = mapDispatchToProps(dispatch)

    const event = {
      preventDefault: () => {},
    }

    received.onDiscoverPokemonClick(event)

    chaiExpect(dispatch).to.have.property('callCount', 1)
  })
})
