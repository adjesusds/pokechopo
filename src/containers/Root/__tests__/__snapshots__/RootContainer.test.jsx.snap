// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RootContainer /> should render the container on valid props 1`] = `
<React.Fragment>
  <Button
    disabled={false}
    onClick={[Function]}
  >
    Discover a Pok√©mon
  </Button>
  <Divider />
  <React.Fragment>
    <Message
      variant="warning"
    >
      Click it.
    </Message>
    <Message
      variant="normal"
    >
      Gotta catch 'em all!
      <Message
        variant="info"
      >
        ‚Ñ¢
      </Message>
    </Message>
  </React.Fragment>
  <Copyright>
    ¬© 
    2018
     
    <Link
      href="https://www.linkedin.com/in/relztic/"
      target="blank"
    >
      Relztic
    </Link>
     
    üê¥
  </Copyright>
</React.Fragment>
`;

exports[`<RootContainer /> should render the container on valid props 2`] = `
<React.Fragment>
  <Button
    disabled={true}
    onClick={[Function]}
  >
    Discover a Pok√©mon
  </Button>
  <Divider />
  <React.Fragment>
    <Message
      variant="warning"
    >
      <Loading
        message="Searching"
        suffix="."
        suffixDuration={1000}
        suffixLength={3}
      />
    </Message>
    <Message
      variant="normal"
    >
      Gotta catch 'em all!
      <Message
        variant="info"
      >
        ‚Ñ¢
      </Message>
    </Message>
  </React.Fragment>
  <Copyright>
    ¬© 
    2018
     
    <Link
      href="https://www.linkedin.com/in/relztic/"
      target="blank"
    >
      Relztic
    </Link>
     
    üê¥
  </Copyright>
</React.Fragment>
`;

exports[`<RootContainer /> should render the container on valid props 3`] = `
<React.Fragment>
  <Button
    disabled={false}
    onClick={[Function]}
  >
    Discover a Pok√©mon
  </Button>
  <Divider />
  <React.Fragment>
    <Message
      variant="warning"
    >
      Click it.
    </Message>
    <Message
      variant="danger"
    >
      InvalidArgumentError: invalid pokemonNumber \`-1\`
    </Message>
  </React.Fragment>
  <Copyright>
    ¬© 
    2018
     
    <Link
      href="https://www.linkedin.com/in/relztic/"
      target="blank"
    >
      Relztic
    </Link>
     
    üê¥
  </Copyright>
</React.Fragment>
`;

exports[`<RootContainer /> should render the container on valid props 4`] = `
<React.Fragment>
  <Button
    disabled={false}
    onClick={[Function]}
  >
    Discover a Pok√©mon
  </Button>
  <Divider />
  <React.Fragment>
    <Message
      variant="normal"
    >
      <Link
        href="https://www.pokemon.com/us/pokedex/bulbasaur/"
        target="blank"
      >
        Bulbasaur
      </Link>
      <Message
        variant="warning"
      >
         #001
      </Message>
    </Message>
    <Message
      variant="success"
    >
      Look at what you got!
    </Message>
  </React.Fragment>
  <Copyright>
    ¬© 
    2018
     
    <Link
      href="https://www.linkedin.com/in/relztic/"
      target="blank"
    >
      Relztic
    </Link>
     
    üê¥
  </Copyright>
</React.Fragment>
`;
